File: codingame/puzzle/hard/genome_sequencing.py
"""
https://www.codingame.com/training/hard/genome-sequencing

============

Input
2
AAC
CCTT

Output
6

============

Input
3
AGATTA
GATTACA
TACAGA

Output
10

============
Input
2
AGATTA
GAT

Output
6
"""
import itertools

n = int(input())
sub_seqs = [input() for i in range(n)]

best = 999

for perms in itertools.permutations(sub_seqs):
    res = ""
    for seq in perms:
        for i in range(0, len(res)+1):
            # check if we can stick new sub-sequence inside existing one
            if i + len(seq) <= len(res):
                if res[i:i+len(seq)] == seq:
                    break
            # check and which point we can add new sub-sequence at the end
            # of existing one. it is always possible because of `0, len(res)+1`-range
            elif res[i:] == seq[:len(res)-i]:
                res += seq[len(res)-i:]
                break
    if len(res) < best:
        best = len(res)

print(best)

File: codingame/puzzle/hard/genome_sequencing.rs
/**
https://www.codingame.com/training/hard/genome-sequencing
*/
use std::io;
use std::cmp;
use itertools::Itertools;

macro_rules! parse_input {
    ($x:expr, $t:ident) => ($x.trim().parse::<$t>().unwrap())
}

fn main() {
    let mut input_line = String::new();
    io::stdin().read_line(&mut input_line).unwrap();
    let n = parse_input!(input_line, usize);

    let sub_seqs = (0..n).map(|_| {
        let mut input_line = String::new();
        io::stdin().read_line(&mut input_line).unwrap();
        let subseq = input_line.trim().to_string();
        subseq
    }).collect::<Vec<String>>();
    let mut best = std::usize::MAX;

    for perms in sub_seqs.into_iter().permutations(n) {
        let mut res = String::new();

        for seq in perms {
            for i in 0..=res.len(){
                let sr = cmp::min(seq.len(), res.len()-i);
                let rr = cmp::min(res.len(), i+seq.len());

                // check if we can stick new sub-sequence inside existing one
                if i + seq.len() <= res.len() {
                    if res[i..rr] == seq {
                        break;
                    }
                } else if i == res.len() || res[i..] == seq[..sr] {
                    // check and which point we can add new sub-sequence at the end
                    // of existing one. it is always possible because of `i == res.len()`
                    res += &seq[res.len()-i..];
                    break;
                }
            }
        }

        if res.len() < best {
            best = res.len()
        }
    }

    println!("{}", best);
}
